sstables_loader.cc

```diff
@@ -213,39 +227,48 @@ future<> sstables_loader::load_and_stream(sstring ks_name, sstring cf_name,
             failed = true;
             eptr = std::current_exception();
             llog.warn("load_and_stream: ops_uuid={}, ks={}, table={}, finish_phase, err={}",
-                    ops_uuid, ks_name, cf_name, eptr);
+                    ops_uuid, s->ks_name(), s->cf_name(), eptr);
         }
         if (!failed) {
             try {
                 co_await coroutine::parallel_for_each(sst_processed, [&] (sstables::shared_sstable& sst) {
                     llog.debug("load_and_stream: ops_uuid={}, ks={}, table={}, remove sst={}",
-                            ops_uuid, ks_name, cf_name, sst->component_filenames());
+                            ops_uuid, s->ks_name(), s->cf_name(), sst->component_filenames());
                     return sst->unlink();
                 });
             } catch (...) {
                 failed = true;
                 eptr = std::current_exception();
                 llog.warn("load_and_stream: ops_uuid={}, ks={}, table={}, del_sst_phase, err={}",
-                        ops_uuid, ks_name, cf_name, eptr);
+                        ops_uuid, s->ks_name(), s->cf_name(), eptr);
             }
         }
         auto duration = std::chrono::duration_cast<std::chrono::duration<float>>(std::chrono::steady_clock::now() - start_time).count();
         for (auto& [node, meta] : metas) {
             llog.info("load_and_stream: ops_uuid={}, ks={}, table={}, target_node={}, num_partitions_sent={}, num_bytes_sent={}",
-                    ops_uuid, ks_name, cf_name, node, meta.num_partitions_sent(), meta.num_bytes_sent());
+                    ops_uuid, s->ks_name(), s->cf_name(), node, meta.num_partitions_sent(), meta.num_bytes_sent());
         }
         auto partition_rate = std::fabs(duration) > FLT_EPSILON ? num_partitions_processed / duration : 0;
         auto bytes_rate = std::fabs(duration) > FLT_EPSILON ? num_bytes_read / duration / 1024 / 1024 : 0;
         auto status = failed ? "failed" : "succeeded";
         llog.info("load_and_stream: finished ops_uuid={}, ks={}, table={}, partitions_processed={} partitions, bytes_processed={} bytes, partitions_per_second={} partitions/s, bytes_per_second={} MiB/s, duration={} s, status={}",
-                ops_uuid, ks_name, cf_name, num_partitions_processed, num_bytes_read, partition_rate, bytes_rate, duration, status);
+                ops_uuid, s->ks_name(), s->cf_name(), num_partitions_processed, num_bytes_read, partition_rate, bytes_rate, duration, status);
         if (failed) {
             std::rethrow_exception(eptr);
         }
     }
     co_return;
 }
 
+future<> sstables_loader::load_and_stream(sstring ks_name, sstring cf_name,
+        ::table_id table_id, std::vector<sstables::shared_sstable> sstables, bool primary_replica_only) {
+    // streamer guarantees topology stability, for correctness, by holding effective_replication_map
+    // throughout its lifetime.
```

Comment:
I think it is safer to take the erm early in the outside layer, .i.e, where we get the ssables lists in sstables_loader::load_new_sstables.  This prevents we get the ssables, then the erm is updated.

Reply:
sstable_directory takes erm during its lifetime too, so we have stability during loading. Taking erm early is not a guarantee migration is not ongoing. It's not a problem if erm is updated before streaming, as long as it won't happen during streaming (which is true as streamer takes erm too). which shard a sstable is assigned to (where we'll stream it from), is not a requirement for correctness.